### PR TITLE
MSAL to use a partioned cache for S2S (client_credential cache)

### PR DESCRIPTION
**Pull request by bgavrilMS**
Benchmark tests: 

|                     Method | TokenCacheSize |         Mean |      Error |     StdDev |
|--------------------------- |--------------- |-------------:|-----------:|-----------:|
| 'Different tenants - O(1)' |           1000 |     60.89 us |   0.660 us |   0.551 us |
|  'Different scopes - O(n)' |           1000 | 44,379.64 us | 415.110 us | 367.984 us |

And a note on the correct calling pattern for multi-tenant apps (@jmprieur): 

```csharp
// this object can be a singleton
var cca = ConfidentialClientApplicationBuilder()
               .WithAuthority("https://login.microsoft.com/common") // we don't know the tenant at this point
              .Build();

// later, when we know the tenant
await cca.AcquireTokenForClient().WithAuthority("https://login.microsoft.com/TENANT_ID").ExecuteAsync();

```

Issue #2544

### CODE REVIEW

File: src/client/Microsoft.Identity.Client/ITokenCacheInternal.cs [hunk 1]
```diff
@@ -71,8 +71,15 @@ internal interface ITokenCacheInternal : ITokenCache, ITokenCacheSerializer
         /// </summary>
         bool HasTokensNoLocks();
 
+
         bool IsTokenCacheSerialized();
 
+        /// <summary>
+        /// MSAL adds serialziation for UWP and also for ConfidentiaClient app token cache.
```
Comment (jennyf19) [inline src/client/Microsoft.Identity.Client/ITokenCacheInternal.cs:78]:
```suggestion
        /// MSAL adds serialization for UWP and also for ConfidentialClient app token cache. 
```

File: src/client/Microsoft.Identity.Client/Cache/CacheImpl/InMemoryPartitionedCacheSerializer.cs [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+            }
+
+            _logger.Verbose($"[InMemoryPartitionedTokenCache] ReadCacheBytes did not find cacheKey {cacheKey}");
+
+            return null;
+        }
+
+        protected override void RemoveKey(string cacheKey)
+        {
+            bool removed = CachePartition.TryRemove(cacheKey, out _);
```
Comment (jennyf19) [inline src/client/Microsoft.Identity.Client/Cache/CacheImpl/InMemoryPartitionedCacheSerializer.cs:39]: is it valuable to include the cache size in all the logs messages, if it's available?

**Review by pmaytak — COMMENTED** [review]
LGTM so far, will review after tests added.

File: src/client/Microsoft.Identity.Client/Cache/CacheImpl/InMemoryPartitionedCacheSerializer.cs (hunk 2 above)
Comment (bgavrilMS) [reply src/client/Microsoft.Identity.Client/Cache/CacheImpl/InMemoryPartitionedCacheSerializer.cs:39]: Added it on Write only, as `Count` is expensive on a ConcurrentDictionary

**Pull request merged**